lib/src/Base/Stat/CovarianceModelImplementation.cxx

```diff
@@ -475,6 +475,209 @@ CovarianceMatrix CovarianceModelImplementation::discretize(const Sample & vertic
+                           const Sample &secondSample,
+                           Matrix &output,
+                           const CovarianceModelImplementation &model)
+      : firstSample_(*firstSample.getImplementation())
+      , secondSample_(*secondSample.getImplementation())
+      , output_(*output.getImplementation())
+      , model_(model)
+  {
+  }
+
+  inline void operator()(const TBB::BlockedRange<UnsignedInteger> &r) const
+  {
+
+    const UnsignedInteger inputDimension = firstSample_.getDimension();
+    for (UnsignedInteger index = r.begin(); index != r.end(); ++index)
+    {
+      // Fill by column
+      // Fill Matrix is firstSampleSize x secondSampleSize
+      // As we might have strong differences between sizes, the loop rely on final blocksize
+      const UnsignedInteger columnIndex = index / firstSample_.getSize();
+      const UnsignedInteger rowIndex = index - columnIndex * firstSample_.getSize();
+      output_(rowIndex, columnIndex) = model_.computeAsScalar(firstSample_.data_begin() + (rowIndex * inputDimension),
+                                                              secondSample_.data_begin() + (columnIndex * inputDimension));
+    }
+  } // operator()
+};
+/* end struct CrossCovarianceFunctor1D */
+
+struct CrossCovarianceFunctor
+{
+  const SampleImplementation &firstSample_;
+  const SampleImplementation &secondSample_;
+  MatrixImplementation &output_;
+  const CovarianceModelImplementation &model_;
+  const UnsignedInteger dimension_;
+
+  CrossCovarianceFunctor(const Sample &firstSample,
+                         const Sample &secondSample,
+                         Matrix &output,
+                         const CovarianceModelImplementation &model)
+      : firstSample_(*firstSample.getImplementation())
+      , secondSample_(*secondSample.getImplementation())
+      , output_(*output.getImplementation())
+      , model_(model)
+      , dimension_(model.getOutputDimension())
+  {
+  }
+
+  inline void operator()(const TBB::BlockedRange<UnsignedInteger> &r) const
+  {
+    for (UnsignedInteger i = r.begin(); i != r.end(); ++i)
+    {
+      // Fill by column
+      // jLocal ==> which column to fill
+      // jBase : use of block size to determine first element of matrix
+      // iLocal : for a fixed jLocal row, which iLocal-th element to fill
+      // iBase : same as jBase but for rows
+      const UnsignedInteger jLocal = i / firstSample_.getSize();
+      const UnsignedInteger jBase = jLocal * dimension_;
+      const UnsignedInteger iLocal = i - jLocal * firstSample_.getSize();
+      const UnsignedInteger iBase = iLocal * dimension_;
+      // Local covariance matrix
+      const SquareMatrix localCovariance(model_(firstSample_[iLocal], secondSample_[jLocal]));
+      for (UnsignedInteger ii = 0; ii < dimension_; ++ii)
+      {
+        for (UnsignedInteger jj = 0; jj < dimension_; ++jj)
+        {
+          output_(iBase + ii, jBase + jj) = localCovariance(ii, jj);
+        }
```

Comment:
you can at least swap the for loops: you want to iterate over contiguous elements in the most inner loop to avoid cache misses as matrices are stores column-wise
(another way of doing this is a loop with std::copy)